### ISSUE TITLE
twilio_account_sub_account auth_token is empty

### ISSUE THREAD

**Issue by enchorb** [RJPearson94/terraform-provider-twilio#52]
## Technical Details

- OS: Windows
- Terraform Version: 1.2.2
- Provider Version 0.17.0

## Affected Resources

- twilio_account_sub_account 

## Describe the bug

The `auth_token` output is always empty

## Steps to Reproduce

Steps to reproduce the behaviour:

A quick way to test and see this is 
```
resource "twilio_account_sub_account" "sub_account" {
  friendly_name = "test"
}

resource "local_file" "twilio" {
  filename = "twilio.json"
  content = jsonencode(twilio_account_sub_account.sub_account)
}
```

**Comment by RJPearson94** [comment]
Hi @enchorb,

Sorry, you have run into this issue. I have been unable to replicate this issue on a mac device, I am trying to get hold of a Windows device to test this again. 

If you look in your Terraform state file, can I check if you see the auth token or if it is also blank in the state file?
Can you provide more information about your terraform code? as the `auth_token` attribute is marked as sensitive and recent terraform versions have tried to improve the handling of sensitive data. So I wonder if this might be related to your issue.

I will continue to see if I can replicate this issue

**Comment by enchorb** [comment]
Thanks for the quick response, I got the same blank response on my Mac as well, even tried wrapping the property with `nonsensitive` 

I posted my full code below in case it helps with debugging. Since I am using your prover and the official Twilio one I labelled yours as `twilio-root`. The credentials passed into the provider are those of the parent account.

```
terraform {
  required_providers {
    twilio = {
      source  = "twilio/twilio"
      version = "0.18.0"
    }

    twilio-root = {
      source = "RJPearson94/twilio"
      version = "0.17.0"
    }
  }
  experiments = [module_variable_optional_attrs]
}

provider "twilio-root" {
  api_key         = var.api_key
  api_secret    = var.api_secret
  account_sid = var.account_sid
}

resource "twilio_account_sub_account" "sub_account" {
  provider           = twilio-root
  friendly_name = var.name
}

resource "local_file" "twilio" {
  filename = "twilio.json"
  content  = jsonencode(twilio_account_sub_account.sub_account)
}

# NOTE: Temporary Blocker As Auth Token Returned Is Empty

provider "twilio" {
  username      = twilio_account_sub_account.sub_account.sid
  password      = twilio_account_sub_account.sub_account.auth_token
  account_sid  = twilio_account_sub_account.sub_account.sid
}

```

**Comment by RJPearson94** [comment]
hmm, this is odd. As I just ran that terraform code locally and it worked. 

Within the Twilio console or via the API, are you seeing the subaccount? i.e. did the provider create it correctly
If can use the API that would be preferred. As you will be able to check if the Twilio API is returning the Auth Token in the response. The docs for retrieving account details are https://www.twilio.com/docs/iam/api/account#fetch-an-account-resource

**Comment by RJPearson94** [comment]
Hi @enchorb, As part of the v0.18.0 release of the provider. I bumped the version of the terraform sdk which included a number of features and bug fixes. While I don’t believe this will solve your issue, would you be able to upgrade to v0.18.0 of the provider and see if you still see the same issue?

**Comment by enchorb** [comment]
Hey thanks for the quick work on the Verify resource! Unfortunately still the same issue for the empty auth token. I do see the subaccount in the console and also if I pass in the accountSid and authToken into the Twilio Node API it is returning the correct response (for both the parent and subaccount)

**Comment by enchorb** [comment]
I do believe for the time being I have a temporary work around and that is using a `null_resource` that calls a custom shell script which returns the auth token from this curl command `curl -X GET 'https://api.twilio.com/2010-04-01/Accounts/ACXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX.json' \
-u $TWILIO_ACCOUNT_SID:$TWILIO_AUTH_TOKEN`

**Comment by RJPearson94** [comment]
hmm, sorry you have to do this workaround. I will continue trying to replicate this issue.

It is currently unclear whether the issue is within the Go SDK or the provider itself. To eliminate the Go SDK, would you be able to run the following [example](https://github.com/RJPearson94/twilio-sdk-go/blob/main/examples/service/api/account/fetch/account_fetch_example.go) substituting in your account details and check that the auth token is present?

Can I check, is the `auth_token` an empty string/ null in the terraform state file or is the attribute missing?

**Comment by enchorb** [comment]
Hey, I will attempt to run that example tomorrow, as for the state file the `auth_token` is empty in there (I shortened the SIDs for security):
```
{
      "module": "module.client.module.twilio",
      "mode": "managed",
      "type": "twilio_account_sub_account",
      "name": "sub_account",
      "provider": "module.client.module.twilio.provider[\"registry.terraform.io/rjpearson94/twilio\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "auth_token": "",
            "date_created": "2022-06-16T19:07:23Z",
            "date_updated": "2022-06-16T19:07:23Z",
            "friendly_name": "flow-qa-pei",
            "id": "AC168b791",
            "owner_account_sid": "ACe7b",
            "sid": "AC168b791",
            "status": "active",
            "timeouts": null,
            "type": "Full"
          },
          "sensitive_attributes": [],
          "private": "eyJlMmJmYjcziY3JlYXRlIjo2MDAwMDAwMDAwMDAsISI6NjAwMDAwMDAwMDAwfX0=",
          "dependencies": [
            "data.sops_file.encrypted"
          ]
        }
      ]
    }
```

**Comment by enchorb** [comment]
Here is the work around that is working, in the .tf file call:
```
data "external" "auth" {
  program  = ["sh", "${path.module}/get_auth_token.sh", var.account_sid, var.auth_token, twilio_account_sub_account.sub_account.sid]
}
```

and in the shell script: 
```
#!/bin/sh

PARENT_SID=$1
PARENT_TOKEN=$2
SUB_SID=$3

main() {
  # Switch To Directory
  cd "$(dirname "$0")" || exit 1

	TOKEN=$(curl -X GET "https://api.twilio.com/2010-04-01/Accounts/$SUB_SID.json" -u "$PARENT_SID:$PARENT_TOKEN" | jq -r .auth_token 2> /dev/null)
	echo "{\"token\": \"$TOKEN\"}"
}

main "$@"
```

**Comment by RJPearson94** [comment]
Thanks for the info @enchorb, I will continue investigating

**Comment by RJPearson94** [comment]
I don't think this will work but trying to eliminate some options, have you tried using the [account details data source](https://registry.terraform.io/providers/RJPearson94/twilio/latest/docs/data-sources/account_details) to lookup the sub-account details? Just wondering if the auth_token attribute is also an empty string

**Comment by RJPearson94** [comment]
Hi @enchorb, I have been able to replicate the issue.

Looking through your terraform code again I see you are initialising the provider with an API Key and Secret, these will be `master` keys within Twilio to allow them to manage the sub-account. I have generated a master API key and created a new sub-account via the provider and through the API, I noticed the `auth_token` is blank in the response from the Twilio API. This is the cause of your issue.

Your script is working because you are authenticating using the Account SID and Auth Token and not the API Key. When using the Account SID and Auth Token, the API returns the auth token for both the current account and any sub-account. 

I have updated the curl command to authenticated using an API Key and Secret, if you run this command you will see the auth token is blank
```
curl -X GET 'https://api.twilio.com/2010-04-01/Accounts/<<ACCOUNT_SID>>.json' \ -u <<API_KEY_SID>>:<<API_KEY_SECRET>>
```

I suspect Twilio does this for security reasons, in the event the API key gets compromised but I cannot find any docs on this, I'm not sure what support you have with Twilio but it might be worth asking the support desk if this is by design. 

In the interim, if you want to access the auth_token of the sub-account you will have to authenticate using the `account_sid` & `auth_token` of the parent account

I hope this helps
